src/command_helpers.rs

```diff
@@ -41,83 +40,122 @@ impl CargoOutput {
-
-        // Capture the standard error coming from compilation, and write it out
-        // with cargo:warning= prefixes. Note that this is a bit wonky to avoid
-        // requiring the output to be UTF-8, we instead just ship bytes from one
-        // location to another.
-        let print = thread::spawn(move || {
-            let mut stderr = BufReader::with_capacity(4096, pipe_reader);
-            let mut line = Vec::with_capacity(20);
-            let stdout = io::stdout();
-
-            // read_until returns 0 on Eof
-            while stderr.read_until(b'\n', &mut line).unwrap() != 0 {
-                {
-                    let mut stdout = stdout.lock();
-
-                    stdout.write_all(b"cargo:warning=").unwrap();
-                    stdout.write_all(&line).unwrap();
-                    stdout.write_all(b"\n").unwrap();
-                }
+const MIN_BUFFER_CAPACITY: usize = 100;
 
-                // read_until does not clear the buffer
-                line.clear();
+impl StderrForwarder {
+    pub(crate) fn new(child: &mut Child) -> Self {
+        Self {
+            inner: child
+                .stderr
+                .take()
+                .map(|stderr| (stderr, Vec::with_capacity(MIN_BUFFER_CAPACITY))),
+            #[cfg(feature = "parallel")]
+            is_non_blocking: false,
+        }
+    }
+
+    fn forward_available(&mut self) -> bool {
+        if let Some((stderr, buffer)) = self.inner.as_mut() {
+            fn write_warning(line: &[u8]) {
+                let stdout = io::stdout();
+                let mut stdout = stdout.lock();
+                stdout.write_all(b"cargo:warning=").unwrap();
+                stdout.write_all(line).unwrap();
+                stdout.write_all(b"\n").unwrap();
             }
-        });
 
-        Ok(Self {
-            handle: Some(print),
-            pipe_writer: Some(pipe_writer),
-        })
+            #[cfg(feature = "parallel")]
+            let read_stderr = if self.is_non_blocking {
+                crate::parallel::stderr::read_non_blocked
+            } else {
+                ChildStderr::read
+            };
+            #[cfg(not(feature = "parallel"))]
+            let read_stderr = ChildStderr::read;
+
+            loop {
+                buffer.reserve(MIN_BUFFER_CAPACITY);
+
+                let old_data_end = buffer.len();
+                buffer.resize(buffer.capacity(), 0);
+                match read_stderr(stderr, &mut buffer[old_data_end..]) {
+                    Err(err) if err.kind() == std::io::ErrorKind::WouldBlock => {
+                        // No data currently, yield back.
+                        buffer.truncate(old_data_end);
+                        return false;
+                    }
+                    Err(err) if err.kind() == std::io::ErrorKind::Interrupted => {
+                        // Interrupted, try again.
+                        buffer.truncate(old_data_end);
```

Comment:
resize is honestly a bit expensive as it have to set the rest of the buffer to 0.

You can either:
 - move the resize outside the loop, and make sure every match returns.
 - use `MaybeUninit`, but given that the `std::io::Read` trait only takes `&[u8]`, it in theory can be a UB if the implementator of `Read` actually reads from it, instead of just writing to it.

Reply:
I switched this to `set_len` since I'm always truncating to the initialized data after reading - I can't move the `resize` outside of the loop (in case we've filled the buffer but haven't seen a `\n`), and I don't think that `MaybeUninit` supports dynamic sized buffers.

Reply:
I was thinking of `Vec<MaybeUninit<u8>>`

Reply:
I don't think that works, `MaybeUninit` isn't guaranteed to be the same layout as the underlying type: https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#layout-1 so you can't get a `&[u8]` from a `Vec<MaybeUninit<u8>>`.

Reply:
Thanjs, I didn't know that.